sonic/ingester.go

```diff
@@ -154,42 +153,55 @@ func (i ingesterChannel) BulkPush(collection, bucket string, parallelRoutines in
 		parallelRoutines = 1
 	}
 
-	errs = make([]IngestBulkError, 0)
-	errMutex := &sync.Mutex{}
-
 	// chunk array into N (parallelRoutines) parts
 	divided := divideIngestBulkRecords(records, parallelRoutines)
 
-	// dispatch each records array into N goroutines
-	group := sync.WaitGroup{}
-	group.Add(len(divided))
+	bulkWriteChan := make(chan []IngestBulkRecord, len(divided))
+	bulkErrorChan := make(chan []IngestBulkError)
+
+	defer close(bulkWriteChan)
+	defer close(bulkErrorChan)
+
 	for _, r := range divided {
-		go func(recs []IngestBulkRecord) {
-			conn, _ := newConnection(i.driver)
-
-			for _, rec := range recs {
-				if conn == nil {
-					addBulkError(&errs, rec, ErrClosed, errMutex)
-				}
-				err := i.Push(collection, bucket, rec.Object, rec.Text)
-				if err != nil {
-					addBulkError(&errs, rec, err, errMutex)
-					continue
-				}
-				// sonic should sent OK
-				_, err = conn.read()
-				if err != nil {
-					addBulkError(&errs, rec, err, errMutex)
-				}
-			}
-			conn.close()
-			group.Done()
-		}(r)
+		bulkWriteChan <- r
+		go bulkPushRoutine(i.driver, collection, bucket, bulkWriteChan, bulkErrorChan)
+	}
+
+	errs = make([]IngestBulkError, 0)
+	for range divided {
+		for _, e := range <-bulkErrorChan {
+			errs = append(errs, e)
+		}
 	}
-	group.Wait()
+
 	return errs
 }
 
+// bulkPushRoutine creates a new ingester instance using the provided driver
+// and pushes all the records received from bulkWriteChan
+// any errors are sent to bulkErrorChan
+func bulkPushRoutine(driver *driver, collection, bucket string, bulkWriteChan <-chan []IngestBulkRecord, bulkErrorChan chan<- []IngestBulkError) {
```

Comment:
In order to be simpler, the `bulkWriteChan <-chan []IngestBulkRecord` parameter could  be replaced with a `[]IngestBulkRecord`. In your case the use of a channel is not really appropriate. You can easily avoid it :)

Reply:
i like that suggestion, will make the change